ouroboros-consensus/src/Ouroboros/Consensus/Node/DbLock.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Ouroboros.Consensus.Node.DbLock (
+    DbLocked (..)
+  , dbLockFile
+  , lockDB
+  ) where
+
+import           Control.Concurrent
+import           Control.Monad (void)
+import           Data.Text (Text)
+import           Ouroboros.Consensus.Storage.FS.API
+import           Ouroboros.Consensus.Storage.FS.API.Types
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Control.Exception
+import           System.FileLock
+
+-- | This can run from multiple processes concurrently, without any locking.
+-- It creates the db directory and its lock file, if any of them doesn't exist
+-- already.
+createDBFolder :: Monad m => HasFS m h -> m ()
+createDBFolder hasFS = do
+    createDirectoryIfMissing hasFS True root
+    void $ hOpen hasFS pFile (WriteMode AllowExisting)
+  where
+    root  = mkFsPath []
+    pFile = fsPathFromList ["dblock"]
+
+-- | We try to lock the db multiple times, before we give up and throw an
+-- exception. Each time, we sleep for an increasing duration. Maximum total
+-- sleeping time is about 1.5 sec. The file is unlocked when the @registry@ is
+-- closed.
+lockDB :: ResourceRegistry IO -> HasFS IO h ->  FilePath -> IO ()
+lockDB registry hasFS dbPath = do
+    createDBFolder hasFS
+    lockAttempt 0
+  where
+    lockAttempt :: Int -> IO ()
+    lockAttempt n = do
+      mlockFile <- allocateEither
+                    registry
+                    (const $ justToRight <$> tryLockFile lockFilePath Exclusive)
+                    (\f -> unlockFile f >> return True)
+      case (mlockFile, n) of
+        (Left (), 5) -> throwIO $ DbLocked lockFilePath
+        (Left (), _) -> do
+          -- TODO: We should log that we failed to take the lock, sleeping and
+          -- retrying.
+          threadDelay $ (n + 1) * 100000
+          lockAttempt $ n + 1
```

Comment:
*If* we go for retries, let's use exponential backoff, not linear. Also we should document the behaviour: first wait x millisec, then ..., upto ..., etc. And we should indeed trace something, but this requires using a `Tracer`, and it's a bit annoying to have to add an extra tracer (`cardano-node` will have to be aware of it) just to trace one message.

Reply:
It looks like the `base` package implementation of file locking on both posix, Linux and Windows will support the blocking lock acquite being interrupted by a timeout (async exception).

This would allow for a simple scheme:
```
res <- timeout 2000000 $ hLock lockfile ExclusiveLock
```
http://hackage.haskell.org/package/base-4.12.0.0/docs/GHC-IO-Handle-Lock.html

And if we're only blocking for up to one or two seconds, we don't need any new tracing I think.
https://github.com/input-output-hk/ouroboros-network/issues/1903#issuecomment-610038489